### DISCUSSION TITLE
Implement Some of the code in Rust

### DISCUSSION THREAD

**Discussion by Ravenant1234** [official-stockfish/Stockfish#4398] [Ideas]
I am newbie developer student and i have read all the benefits of Rust has to offer against C/C++ in memory safety and i think the benefit is so good even Linus Torvalds is using rust for rewriting some parts of the Linux kernel (The Rust For Linux ojeda project) (Wich i think is more performance sensitive that a chess engine)
I have read also that rust and c++ has very good interoperability

**Comment by Ravenant1234** [comment]
Maybe a Rust re-implementation of some parts of the engine may help to reduce memory bugs, performance issues and speed up the development process.
There is also a various partial implementation of the engine in rust, for example:
https://github.com/stouset/stockfish-rs

**Reply by TheBlackPlague** [reply]
> Maybe a Rust re-implementation of some parts of the engine may help to reduce memory bugs, performance issues and speed up the development process. There is also a various partial implementation of the engine in rust, for example: https://github.com/stouset/stockfish-rs

What are the memory bugs in Stockfish that you speak of?

**Reply by Disservin** [reply]
also porting everything to rust wouldnt speedup the development process at all and personally i find rust syntax much more difficult to read lol and re implementating only some parts is quite weird if you ask me

**Reply by Ravenant1234** [reply]
> > Maybe a Rust re-implementation of some parts of the engine may help to reduce memory bugs, performance issues and speed up the development process. There is also a various partial implementation of the engine in rust, for example: https://github.com/stouset/stockfish-rs
> 
> What are the memory bugs in Stockfish that you speak of?

Memory bugs in c++ itself, there are a Lot of info about that. Search "C++ vs Rust". And Rust have an superb compiler (it Does The most of The job for you, Is really awesome)

**Reply by Ravenant1234** [reply]
Linus Is rewriting The Linux kernel in Rust wich Is Extremely Performance AND memory sensitive. Search "Rust for Linux" AND u Will see

**Reply by Ravenant1234** [reply]
You didnt need to rewrite ALL of The code in RUST from zero. Just start from here: https://github.com/pleco-rs/Pleco That Is a rewrite of The Stockfish engine.

**Reply by Ravenant1234** [reply]
> You didnt need to rewrite ALL of The code in RUST from zero. Just start from here: https://github.com/pleco-rs/Pleco That Is a rewrite of The Stockfish engine. 

There are endless guides for porting c++ to rust

**Reply by Ravenant1234** [reply]
> also porting everything to rust wouldnt speedup the development process at all and personally i find rust syntax much more difficult to read lol and re implementating only some parts is quite weird if you ask me

More difficult to read? The only difficult syntax part i Have encountered (i am a begginer in RUST) Is The Procedural Macros syntax.

**Reply by TheBlackPlague** [reply]
> > > Maybe a Rust re-implementation of some parts of the engine may help to reduce memory bugs, performance issues and speed up the development process. There is also a various partial implementation of the engine in rust, for example: https://github.com/stouset/stockfish-rs
> > 
> > 
> > What are the memory bugs in Stockfish that you speak of?
> 
> Memory bugs in c++ itself, there are a Lot of info about that. Search "C++ vs Rust". And Rust have an superb compiler (it Does The most of The job for you, Is really awesome)

Clearly, you have no clue about anything you're talking about.

**Reply by TheBlackPlague** [reply]
As I stated earlier, you don't know what you're saying. You go off the windmill that "Rust = good, memory-safe, C++ = memory bugs" - this isn't true. And a foolish notion, at that.

C++ code is inherently unsafe. It allows for bugs to happen, whereas Rust's compiler and/or the language standard doesn't. That's the only difference. Whether bugs exist is a different story, and there's zero indication why we can't fix them directly in C++ code. If you have a specific bug in mind, state it. Otherwise, please think before you post stupid statements on the internet. 

*All unsafe code is safe when used safely.*
